models/llama4/generation.py

```diff
@@ -82,19 +82,39 @@ def build(
 
         state_dict = maybe_reshard_state_dict(
             ckpt_paths,
-            n_kv_heads=model_args.n_kv_heads if model_args.n_kv_heads else model_args.n_heads,
+            n_kv_heads=(model_args.n_kv_heads if model_args.n_kv_heads else model_args.n_heads),
             moe_num_experts=model_args.moe_args.num_experts,
         )
         print("Loaded checkpoint")
         if quantization_mode == QuantizationMode.fp8_mixed or quantization_mode == QuantizationMode.int4_mixed:
             from .quantization.loader import convert_to_quantized_model
 
+            scale_state_dict = None
+            if quantization_mode == QuantizationMode.int4_mixed:
+                scale_ckpt_paths = sorted(Path(ckpt_dir).glob("*.pt"))
+
+                if len(scale_ckpt_paths) > 0:
+                    print(f"Loading a scale checkpoint (shards={len(scale_ckpt_paths)}, current-mp-size={world_size})")
+                    scale_state_dict = maybe_reshard_state_dict(
+                        scale_ckpt_paths,
+                        n_kv_heads=(model_args.n_kv_heads if model_args.n_kv_heads else model_args.n_heads),
+                        moe_num_experts=model_args.moe_args.num_experts,
+                        is_scale=True,
+                    )
+                    model_args.quantization_args = QuantizationArgs()
+                    model_args.quantization_args.int4_weight = True
+                    print("Loaded scale checkpoint")
             torch.set_default_tensor_type(torch.BFloat16Tensor)
             model = Transformer(model_args)
             print("Loading state dict...")
             model.load_state_dict(state_dict, strict=False)
```

Comment:
if you move the `model.load_state_dict()` to `convert_to_quantized_model()` then you can do the following:

- change the structure of the Transformer from the outside in this code path (whatever you are doing with Experts)
- move all this scale ckpt paths complexity into quantization land

nobody reading generation.py should know about quantization unless they want to dig into it.